### PR TITLE
[AMORO-2378] The optimizer based on Flink on YARN should prioritize loading the optimizer-job.jar

### LINKED ISSUE #2378: [Improvement]: The optimizer based on Flink on YARN should prioritize loading the optimizer-job.jar
### Search before asking

- [X] I have searched in the [issues](https://github.com/NetEase/amoro/issues?q=is%3Aissue) and found no similar issues.

### What would you like to be improved?

When there are different versions of the iceberg package (or other packages involving serialization and deserialization) present in the runtime environment, it can result in the failure of object deserialization.

d8a83e9994aa75939283a566d51946f

### How should we improve?

 We should prioritize loading the optimizer-job.jar (a fat jar) to avoid being affected by different versions of classes from other packages.

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Subtasks

_No response_

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by hameizi**
## Why are the changes needed?

Close #2378.

## Brief change log
add config `yarn.per-job-cluster.include-user-jar:FIRST`

## How was this patch tested?

- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] Run test locally before making a pull request

## Documentation

- Does this pull request introduce a new feature? (no)
- If yes, how is the feature documented? (not documented)

### CODE REVIEW

File: ams/server/src/main/java/com/netease/arctic/server/manager/FlinkOptimizerContainer.java [hunk 1]
```diff
@@ -416,6 +420,8 @@ public static class FlinkConfKeys {
     public static final String JOB_MANAGER_TOTAL_PROCESS_MEMORY = "jobmanager.memory.process.size";
     public static final String TASK_MANAGER_TOTAL_PROCESS_MEMORY =
         "taskmanager.memory.process.size";
+    public static final String USER_LIB_FIRST = "yarn.per-job-cluster.include-user-jar";
```
Comment (wangtaohz) [inline ams/server/src/main/java/com/netease/arctic/server/manager/FlinkOptimizerContainer.java:423]: `USER_LIB_FIRST` may not be an approprite name. How about using the same property name [CLASSPATH_INCLUDE_USER_JAR](https://github.com/apache/flink/blob/d5b1afb7f67fe7da166193deb7711b8a8b163bcf/flink-yarn/src/main/java/org/apache/flink/yarn/configuration/YarnConfigOptions.java#L57C1-L57C1) from Flink?

File: ams/server/src/main/java/com/netease/arctic/server/manager/FlinkOptimizerContainer.java [hunk 2]
```diff
@@ -226,6 +228,12 @@ private void addKubernetesProperties(Resource resource, FlinkConf flinkConf) {
     flinkConf.putToOptions(FlinkConfKeys.KUBERNETES_JOBMANAGER_LABLES, resourceLabel);
   }
 
+  private void addYarnProperties(FlinkConf flinkConf) {
+    // load optimizer jar first
```
Comment (wangtaohz) [inline ams/server/src/main/java/com/netease/arctic/server/manager/FlinkOptimizerContainer.java:232]: Nit: Capitalize the first letter of comments.

File: ams/server/src/main/java/com/netease/arctic/server/manager/FlinkOptimizerContainer.java [hunk 3]
```diff
@@ -183,9 +183,15 @@ protected String buildOptimizerStartupArgsString(Resource resource) {
     resourceFlinkConf.putToOptions(
         FlinkConfKeys.TASK_MANAGER_TOTAL_PROCESS_MEMORY, taskManagerMemory + "m");
 
+    // Set memory managed fraction
+    resourceFlinkConf.putToOptions(
+        FlinkConfKeys.MANAGED_MEMORY_FRACTION, FlinkConfKeys.MANAGED_MEMORY_FRACTION_DEFAULT);
```
Comment (wangtaohz) [inline ams/server/src/main/java/com/netease/arctic/server/manager/FlinkOptimizerContainer.java:188]: We'd better not include irrelevant changes in the PR. 

Also, in the docs, we recommend this configuration [flink-conf.taskmanager.memory.managed.size=32mb](https://github.com/NetEase/amoro/blob/ca2061edcb117b472eb9e0a3c84933c37118b79e/docs/admin-guides/managing-optimizers.md?plain=1#L56)

**Review by zhoujinsong — APPROVED** [review]
LGTM.

**Review by wangtaohz — APPROVED** [review]
LGTM

**Pull request merged**